x-pack/auditbeat/module/system/package/package.go

```diff
@@ -0,0 +1,537 @@
+	h.WriteString(pkg.Name)
+	h.WriteString(pkg.InstallTime.String())
+	return h.Sum64()
+}
+
+func (pkg Package) toMapStr() common.MapStr {
+	mapstr := common.MapStr{
+		"name":        pkg.Name,
+		"version":     pkg.Version,
+		"installtime": pkg.InstallTime,
+	}
+
+	if pkg.Arch != "" {
+		mapstr.Put("arch", pkg.Arch)
+	}
+
+	if pkg.License != "" {
+		mapstr.Put("license", pkg.License)
+	}
+
+	if pkg.Release != "" {
+		mapstr.Put("release", pkg.Release)
+	}
+
+	if pkg.Size != 0 {
+		mapstr.Put("size", pkg.Size)
+	}
+
+	if pkg.Summary != "" {
+		mapstr.Put("summary", pkg.Summary)
+	}
+
+	if pkg.URL != "" {
+		mapstr.Put("url", pkg.URL)
+	}
+
+	return mapstr
+}
+
+func getOS() (*types.OSInfo, error) {
+	host, err := sysinfo.Host()
+	if err != nil {
+		return nil, errors.Wrap(err, "error getting the OS")
+	}
+
+	hostInfo := host.Info()
+	if hostInfo.OS == nil {
+		return nil, errors.New("no host info")
+	}
+
+	return hostInfo.OS, nil
+}
+
+// New constructs a new MetricSet.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	cfgwarn.Experimental("The %v/%v dataset is experimental", moduleName, metricsetName)
+
+	config := defaultConfig()
+	if err := base.Module().UnpackConfig(&config); err != nil {
+		return nil, errors.Wrapf(err, "failed to unpack the %v/%v config", moduleName, metricsetName)
+	}
+
+	bucket, err := datastore.OpenBucket(bucketName)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to open persistent datastore")
+	}
+
+	ms := &MetricSet{
+		BaseMetricSet: base,
+		config:        config,
+		log:           logp.NewLogger(metricsetName),
+		cache:         cache.New(),
+		bucket:        bucket,
+	}
+
+	osInfo, err := getOS()
+	if err != nil {
+		return nil, errors.Wrap(err, "error determining operating system")
+	}
+	ms.osFamily = osInfo.Family
+	switch osInfo.Family {
+	case redhat:
+		return nil, fmt.Errorf("RPM support is not yet implemented")
+	case debian:
+		if _, err := os.Stat(dpkgStatusFile); err != nil {
+			return nil, errors.Wrapf(err, "error looking up %s", dpkgStatusFile)
+		}
+	case darwin:
+		if _, err := os.Stat(homebrewCellarPath); err != nil {
+			return nil, errors.Wrapf(err, "error looking up %s - is Homebrew installed?", homebrewCellarPath)
+		}
+	default:
+		return nil, fmt.Errorf("this metricset does not support OS family %v", osInfo.Family)
+	}
+
+	// Load from disk: Time when state was last sent
+	err = bucket.Load(bucketKeyStateTimestamp, func(blob []byte) error {
+		if len(blob) > 0 {
+			return ms.lastState.UnmarshalBinary(blob)
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	if !ms.lastState.IsZero() {
+		ms.log.Debugf("Last state was sent at %v. Next state update by %v.", ms.lastState, ms.lastState.Add(ms.config.effectiveStatePeriod()))
+	} else {
+		ms.log.Debug("No state timestamp found")
+	}
+
+	// Load from disk: Packages
+	packages, err := ms.restorePackagesFromDisk()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to restore packages from disk")
```

Comment:
I think that returning an error here is going to cause Auditbeat to stop, right? Not sure if we discussed this before, but I wonder if recovering from this condition is not something that we want, to protect from corrupted gob.

We could print a stacktrace to the logs with a clear "please report this" message, but still continue the process.

I'm not convinced that we should recovered from the error, but I think it's worth discussing.

Reply:
> I think that returning an error here is going to cause Auditbeat to stop, right?

Yes. If the `beat.db` file is corrupted, the user will have to delete it to start.

My first instinct is to be conservative and surface the error. Something unexpected happened - maybe innocent, maybe not, we don't know - and the user should address it before continuing.

One alternative would be to print a warning but continue, e.g. treating the file as non existent which would trigger a whole state update to be sent. But that would miss any changes made while Auditbeat was down.

I think my opinion at this time is that unless we have a somewhat common case in which this error can be triggered and there is a good reason to not stop in that case I would be in favor of stopping as we do now.

Reply:
Ok, I agree. Let's start with stopping Auditbeat in this case.

My concern is that a break in compatibility in `beat.db` is going to escalate itself to blocker bugs (Auditbeat doesn't start at all). In order to protect against this, we should have tests with some `beat.db` golden files from older versions.